src/onewire.rs

```diff
@@ -0,0 +1,102 @@
+use core::marker::PhantomData;
+use core::ptr;
+
+use esp_idf_sys::*;
+
+use crate::peripheral::Peripheral;
+
+#[derive(Debug)]
+pub struct Device<'b> {
+    _device: onewire::onewire_device_t,
+    // _bus: &'b BusDriver<'b>, // not sure how I would hold a reference to this aside from the handle
```

Comment:
@Vollbrecht
The internal esp-idf struct for `onewire_device_t` holds a handle to the onewire bus, and in rusty terms it means the `Device` lifetime is linked to the lifetime of the bus.

Does it make sense to include a reference to the bus wrapper so that the user can call `Device::bus(&self)` and get a `&BusDriver` return value? Or is it sufficient to just return the internal handle type?

Reply:
@ivmarkov any input?

Reply:
Look at `SpiDeviceDriver`. Ideally, you need a _non_mutable_ borrow of the bus driver to be kept into `OWDeviceDriver`. For maximum flexibility (where the user either borrows a ref to the bus driver, or an `Arc` ref, or just moves an owned instance of the bus driver, use `T: Borrow<OWBusDriver>` - again - see `SpiDeviceDriver`.

Reply:
But yes, you DO need to track the lifetime of the bus driver somehow (ideally- with a `Borrow` generic as per above). Else nothing stops the user from dropping the bus driver while the device driver is still alive, which is not ideal, to put it mildly.

Reply:
I attempted this, and it got pretty gnarly pretty quickly. https://gist.github.com/DaneSlattery/e0236dac98224bfe6f0137d75e209ca4

I don't know how to enforce that the device is dropped before the search is dropped, and my example does not compile anymore. I need to be able to drop a device without dropping the search.
```
error[E0597]: `search` does not live long enough
  --> examples/rmt_onewire.rs:68:23
   |
67 |     let mut search: DeviceSearch<OneWireBusDriver> = rmt_onewire.search()?;
   |         ---------- binding `search` declared here
68 |     let next_device = search.next_device();
   |                       ^^^^^^ borrowed value does not live long enough
...
80 | }
   | -
   | |
   | `search` dropped here while still borrowed
   | borrow might be used here, when `search` is dropped and runs the `Drop` code for type `DeviceSearch`
```

Reply:
This taught me a fair amount about lifetimes, thank you. I have implemented most of your suggestions. 

If I remove the phantomdata, I get this error about the lifetimes in `DeviceSearch` and `OWDevice`, even though the lifetime is captured in the constraint `T: Borrow<OWDriver<'b>>`

```
error[E0392]: lifetime parameter `'b` is never used
  --> src/onewire.rs:59:25
   |
59 | pub struct DeviceSearch<'b, T>
   |                         ^^ unused lifetime parameter
   |
   = help: consider removing `'b`, referring to it in a field, or using a marker such as `core::marker::PhantomData`

```

Reply:
Given my comments below that `DeviceSearch` should actually NOT use a `T: Borrow<OWDevice<'b>>` (and by the way **sorry** if I wrongly seemed to hint you the other way in my comments above!!) - and given that `OWDevice` is likely no more, I'll resolve this.

As for your concrete problem at hand (if you face it in _other_ circumstances) - wouldn't the following work:
```rust
pub struct DeviceSearch<T> { ... }  // I.e. NO lifetime `'b` on the `DeviceSearch` struct itself

impl<'b, T> DeviceSearch<T>
where 
    T: Borrow<OWDevice<'b>>,
{
...
}
```
(Need to try that myself though, but don't see why it wouldn't work.)
?